pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/MissingOverrideRule.java

```
@@ -0,0 +1,363 @@
+ * @since 6.1.0
+ */
+public class MissingOverrideRule extends AbstractJavaRule {
+
+    private final Stack<MethodLookup> currentLookup = new Stack<>();
+
+
+    @Override
+    public Object visit(ASTCompilationUnit node, Object data) {
+        currentLookup.clear();
+        return super.visit(node, data);
+    }
+
+
+    @Override
+    public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
+        currentLookup.push(getMethodLookup(node.getType()));
+        super.visit(node, data);
+        currentLookup.pop();
+
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTEnumDeclaration node, Object data) {
+        currentLookup.push(getMethodLookup(node.getType()));
+        super.visit(node, data);
+        currentLookup.pop();
+
+        return data;
+    }
+
+
+
+    @Override
+    public Object visit(ASTAllocationExpression node, Object data) {
+        if (node.isAnonymousClass()) {
+            currentLookup.push(getMethodLookup(node.getType()));
+        }
+        super.visit(node, data);
+
+        if (node.isAnonymousClass()) {
+            currentLookup.pop();
+        }
+
+        return data;
+    }
+
+
+    @Override
+    public Object visit(ASTEnumConstant node, Object data) {
+        // FIXME, ASTEnumConstant needs typeres support!
+        //        if (node.isAnonymousClass()) {
+        //            currentExploredClass.push(node.getType());
+        //        }
+        super.visit(node, data);
+
+        //        if (node.isAnonymousClass()) {
+        //            currentExploredClass.pop();
+        //        }
+
+        return data;
+    }
+
+
+    /**
+     * Returns a map of method name to methods with the same name (overloads).
+     * The map contains a MethodWrapper for all methods declared in this class.
+     *
+     * @param exploredType Type to explore
+     */
+    private MethodLookup getMethodLookup(Class<?> exploredType) {
+        if (exploredType == null) {
+            return null;
```

If we return here an "empty" MethodLookup instance, we could avoid the null check for currentLookup.peek() later on.

---

Yes, but that null check allows to fail early, before we checked all annotations and evaluated the arguments to `isOverridden`. The performance loss would probably be insignificant, but the null check, if it's written explicitly, also documents that the lookup can be null/ empty/ unresolved, whereas using an empty instance would hide it I think